interpreter/prune.go

```diff
@@ -0,0 +1,126 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package interpreter
+
+import (
+	"fmt"
+	"github.com/golang/protobuf/ptypes/struct"
+	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/common/types/ref"
+	expr "github.com/google/cel-spec/proto/v1/syntax"
+)
+
+type exprPruner struct {
+	expr    *expr.Expr
+	program Program
+	state   EvalState
+}
+
+func PruneExpr(expr *expr.Expr, program Program, state EvalState) {
+	pruner := &exprPruner{
+		expr:    expr,
+		program: program,
+		state:   state}
+	pruner.prune(expr)
+}
+
+func ConstantFoldExpr(expr *expr.ParsedExpr, interpreter Interpreter, container string) interface{} {
+	program := NewProgram(expr.Expr, expr.SourceInfo, container)
+	interpretable := interpreter.NewInterpretable(program)
+	result, state := interpretable.Eval(
+		NewActivation(map[string]interface{}{}))
+	// We could prune even in case of error but that is probably not very useful.
+	if !types.IsError(result) {
+		PruneExpr(expr.Expr, program, state)
+	}
+	return result
```

Comment:
I think you'll want to preserve the original expression and build a new one since there could be other inputs provided for eval against the expression at a later date.

Reply:
I think both for PruneExpr and ConstantFoldExpr interfaces, it is better to leave potential cloning to the caller. Cloning might not be necessary or might be necessary in a subset of calls.